controllers/operatorconfigmap_controller.go

```diff
@@ -0,0 +1,484 @@
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update cephfs deployment")
+			return ctrl.Result{}, err
+		}
+
+		err = c.createOrUpdate(c.cephFSDaemonSet, func() error {
+			c.cephFSDaemonSet = csi.GetCephFSDaemonSet(c.OperatorNamespace)
+			if err := c.own(c.cephFSDaemonSet, OperatorConfigMap); err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update cephfs daemonset")
+			return ctrl.Result{}, err
+		}
+
+		err = c.createOrUpdate(c.rbdDeployment, func() error {
+			c.rbdDeployment = csi.GetRBDDeployment(c.OperatorNamespace)
+			if err := c.own(c.rbdDeployment, OperatorConfigMap); err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update rbd deployment")
+			return ctrl.Result{}, err
+		}
+
+		err = c.createOrUpdate(c.rbdDaemonSet, func() error {
+			c.rbdDaemonSet = csi.GetRBDDaemonSet(c.OperatorNamespace)
+			if err := c.own(c.rbdDaemonSet, OperatorConfigMap); err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update rbd daemonset")
+			return ctrl.Result{}, err
+		}
+
+		// Need to handle deletion of the csiDriver object, we cannot set
+		// ownerReference on it as its cluster scoped resource
+		cephfsCSIDriver := templates.CephFSCSIDriver.DeepCopy()
+		cephfsCSIDriver.ObjectMeta.Name = csi.GetCephFSDriverName()
+		if err := csi.CreateCSIDriver(c.ctx, c.Client, cephfsCSIDriver); err != nil {
+			c.log.Error(err, "unable to create cephfs CSIDriver")
+			return ctrl.Result{}, err
+		}
+
+		rbdCSIDriver := templates.RbdCSIDriver.DeepCopy()
+		rbdCSIDriver.ObjectMeta.Name = csi.GetRBDDriverName()
+		if err := csi.CreateCSIDriver(c.ctx, c.Client, rbdCSIDriver); err != nil {
+			c.log.Error(err, "unable to create rbd CSIDriver")
+			return ctrl.Result{}, err
+		}
+
+		prometheusRule := &monitoringv1.PrometheusRule{}
+		if err := k8sYAML.NewYAMLOrJSONDecoder(bytes.NewBufferString(string(pvcPrometheusRules)), 1000).Decode(prometheusRule); err != nil {
+			c.log.Error(err, "Unable to retrieve prometheus rules.", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+			return ctrl.Result{}, err
+		}
+
+		prometheusRule.SetNamespace(c.OperatorNamespace)
+
+		err = c.createOrUpdate(prometheusRule, func() error {
+			applyLabels(OperatorConfigMap.Data["OCS_METRICS_LABELS"], &prometheusRule.ObjectMeta)
+			return c.own(prometheusRule, OperatorConfigMap)
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update prometheus rules")
+			return ctrl.Result{}, err
+		}
+
+		c.log.Info("prometheus rules deployed", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+	}
+	return ctrl.Result{}, nil
+}
+
+func (c *OperatorConfigMapReconciler) deletionPhase() error {
+	if err := c.pvsPresent(); err != nil {
```

Comment:
we aren't interested on all pvs as the func name suggests, `hasOCSVolumes` isn't a bad name.